core/templates/dev/head/domain/exploration/EditableExplorationBackendApiService.ts

```diff
@@ -18,108 +18,123 @@
+              var exploration = angular.copy(response.data);
+
+              // Delete from the ReadOnlyExplorationBackendApiService's cache
+              // As the two versions of the data (learner and editor) now differ
+              ReadOnlyExplorationBackendApiService.deleteExplorationFromCache(
+                explorationId, exploration);
+
+              if (successCallback) {
+                successCallback(exploration);
+              }
+            }, function(errorResponse) {
+              if (errorCallback) {
+                errorCallback(errorResponse.data);
+              }
+            }
+          );
         }
-      }, function(errorResponse) {
-        if (errorCallback) {
-          errorCallback(errorResponse.data);
-        }
-      });
+      );
     };
 
     var _deleteExploration = function(
         explorationId, successCallback, errorCallback) {
-      var editableExplorationDataUrl = _getExplorationUrl(explorationId, null);
-
-      $http['delete'](editableExplorationDataUrl).then(function() {
-        // Delete item from the ReadOnlyExplorationBackendApiService's cache
-        ReadOnlyExplorationBackendApiService.deleteExplorationFromCache(
-          explorationId);
-        if (successCallback) {
-          successCallback({});
-        }
-      }, function(errorResponse) {
-        if (errorCallback) {
-          errorCallback(errorResponse.data);
+      _getExplorationUrl(explorationId, null).then(
+        function(editableExplorationDataUrl) {
+          $http['delete'](editableExplorationDataUrl).then(function() {
+            // Delete item from the ReadOnlyExplorationBackendApiService's cache
+            ReadOnlyExplorationBackendApiService.deleteExplorationFromCache(
+              explorationId);
+            if (successCallback) {
+              successCallback({});
+            }
+          }, function(errorResponse) {
+            if (errorCallback) {
+              errorCallback(errorResponse.data);
+            }
+          });
         }
-      });
+      );
     };
 
     var _getExplorationUrl = function(explorationId, applyDraft) {
       if (applyDraft) {
-        return UrlInterpolationService.interpolateUrl(
+        return $q.resolve(UrlInterpolationService.interpolateUrl(
           EDITABLE_EXPLORATION_DATA_DRAFT_URL_TEMPLATE, {
             exploration_id: explorationId,
             apply_draft: JSON.stringify(applyDraft)
-          });
-      }
-      if (!GLOBALS.can_edit && GLOBALS.can_voiceover) {
-        return UrlInterpolationService.interpolateUrl(
-          VOICEOVER_EXPLORATION_DATA_URL_TEMPLATE, {
-            exploration_id: explorationId
-          });
+          })
+        );
       }
-
-      return UrlInterpolationService.interpolateUrl(
-        EDITABLE_EXPLORATION_DATA_URL_TEMPLATE, {
-          exploration_id: explorationId
-        });
+      return ExplorationRightsDataService.getRightsAsync().then(
```

Comment:
I think this is making it complicated, can we have a service which will initialize itself after calling  DataSertvice and we can call this new services to check whether the user can edit/voiceover exploration? (UserExplorationRightsService.canUserEditExploration())

I don't have a good suggestion here but it seems complicated to me, @vojtechjelinek do you have any suggestion here?